xapian-core/weight/lmweight.cc

```diff
@@ -200,16 +222,34 @@ LMWeight::get_maxpart() const
     return (upper_bound * param_log > 1.0) ? log(upper_bound * param_log) : 1.0;
 }
 
+
+/* The extra weight component in the Dir+ formula is :-
+ *
+ * |Q| * log (param_smoothing1 / (|D| + param_smoothing1))
+ * 
+ * where, |Q| is total query length.
+ *	  |D| is total document length.
+ */
 double
-LMWeight::get_sumextra(Xapian::termcount, Xapian::termcount) const
+LMWeight::get_sumextra(Xapian::termcount len, Xapian::termcount) const
 {
-    return 0;
+    if (select_smoothing == DIRICHLET_PLUS_SMOOTHING) {
+	double extra_weight = param_smoothing1 / (len + param_smoothing1);
+	return get_query_length() * log(extra_weight);
+    }
+    else
+	return 0;
 }
 
 double
 LMWeight::get_maxextra() const
 {
-    return 0;
+    if (select_smoothing == DIRICHLET_PLUS_SMOOTHING) {
+	double extra_weight = param_smoothing1 / (get_doclength_lower_bound() + param_smoothing1);
```

Comment:
You haven't requested the statistic used here (i.e. a lower bound on the document length), so that is presumably the uninitialised value - you need to add something like this to the constructor:

```
if (select_smoothing == DIRICHLET_PLUS_SMOOTHING)
   need_stat(DOC_LENGTH_MIN);
```

More generally, you can run `valgrind --track-origins=yes` to find the origin of uninitialised values (it's slower to do so, which is why it's not the default), e.g.:

```
./runtest valgrind  --track-origins=yes ./apitest -b glass unigramlmweight7
```